crates/catalog/sql/src/catalog.rs

```diff
@@ -0,0 +1,517 @@
+            .connect(&config.url)
+            .await
+            .map_err(from_sqlx_error)?;
+
+        sqlx::query(
+            &("create table if not exists ".to_string()
+                + CATALOG_TABLE_VIEW_NAME
+                + " ("
+                + CATALOG_NAME
+                + " varchar(255) not null,"
+                + TABLE_NAMESPACE
+                + " varchar(255) not null,"
+                + TABLE_NAME
+                + " varchar(255) not null,"
+                + METADATA_LOCATION_PROP
+                + " varchar(255),"
+                + PREVIOUS_METADATA_LOCATION_PROP
+                + " varchar(255),"
+                + RECORD_TYPE
+                + " varchar(5), primary key ("
+                + CATALOG_NAME
+                + ", "
+                + TABLE_NAMESPACE
+                + ", "
+                + TABLE_NAME
+                + ")
+                        );"),
+        )
+        .execute(&pool)
+        .await
+        .map_err(from_sqlx_error)?;
+
+        sqlx::query(
+            "create table if not exists iceberg_namespace_properties (
+                            catalog_name varchar(255) not null,
+                            namespace varchar(255) not null,
+                            property_key varchar(255),
+                            property_value varchar(255),
+                            primary key (catalog_name, namespace, property_key)
+                        );",
+        )
+        .execute(&pool)
+        .await
+        .map_err(from_sqlx_error)?;
+
+        let file_io = FileIO::from_path(&config.warehouse)?
+            .with_props(&config.props)
+            .build()?;
+
+        Ok(SqlCatalog {
+            name: config.name.to_owned(),
+            connection: pool,
+            storage: file_io,
+            cache: Arc::new(DashMap::new()),
+        })
+    }
+}
+
+#[derive(Debug)]
+struct TableRef {
+    table_namespace: String,
+    table_name: String,
+    metadata_location: String,
+    _previous_metadata_location: Option<String>,
+}
+
+fn query_map(row: &AnyRow) -> std::result::Result<TableRef, sqlx::Error> {
+    Ok(TableRef {
+        table_namespace: row.try_get(0)?,
+        table_name: row.try_get(1)?,
+        metadata_location: row.try_get(2)?,
+        _previous_metadata_location: row.try_get::<String, _>(3).map(Some).or_else(|err| {
+            if let sqlx::Error::ColumnDecode {
+                index: _,
+                source: _,
+            } = err
+            {
+                Ok(None)
+            } else {
+                Err(err)
+            }
+        })?,
+    })
+}
+
+#[async_trait]
+impl Catalog for SqlCatalog {
+    async fn list_namespaces(
+        &self,
+        _parent: Option<&NamespaceIdent>,
+    ) -> Result<Vec<NamespaceIdent>> {
+        let name = self.name.clone();
+        let rows = sqlx::query(
+            "select distinct table_namespace from iceberg_tables where catalog_name = ?;",
```

Comment:
This seems incorrect to me, if a namespace has no table or view, it's missing here. We should query namespace table?